Update network-runner docs (#719)
ava-labs/avalanche-docs: docs/quickstart/network-runner.md

# Avalanche Network Runner

The Avalanche Network Runner **(ANR)** allows a user to define, create and interact with a network of Avalanche nodes. It can be used for development and testing.

Developing P2P systems is hard, and blockchains are no different. A developer can't just focus on the functionality of a node, but needs to consider the dynamics of the network, the interaction of nodes and emergent system properties. A lot of testing can't be addressed by unit testing, but needs a special kind of integration testing, where the code runs in interaction with other nodes, attempting to simulate real network scenarios.

> In the context of avalanche, **[subnets](https://docs.avax.network/build/tutorials/platform/subnets/)** are a special focus which requires new tooling and support for playing, working and testing with this unique feature of the Avalanche ecosystem. 

Reviewer: ```suggestion
In the context of avalanche, **[subnets](../subnets/README.md)** are a special focus which requires new tooling and support for playing, working and testing with this unique feature of the Avalanche ecosystem. 
```

Reviewer: We should use relative path to reference other docs.

Clone the repository with:

```bash
git clone https://github.com/ava-labs/avalanche-network-runner.git
```

> There are also binary releases ready to use at [releases](https://github.com/ava-labs/avalanche-network-runner/releases) ready to use.

Reviewer: ```suggestion
There are also binary releases ready to use at [releases](https://github.com/ava-labs/avalanche-network-runner/releases).
```

To build and install the binary locally (requires `golang` to be installed. Check the [requirements](https://github.com/ava-labs/avalanchego#installation) for the minimum version):
```bash
cd ${HOME}/go/src/github.com/ava-labs/avalanche-network-runner
go install -v ./cmd/avalanche-network-runner
```

Unless otherwise specified, file paths given below are relative to the root of this repository. 

## Usage

There are two main ways to use the network-runner:

> * Run ANR as a binary

Reviewer: add a new line. Otherwise, there is no line break on the doc site.

  This is the recommended approach for most use cases. Doesn't require golang installation and provides a RPC server with an HTTP API and a client library for easy interaction.
* Import this repository into your go program

>   This allows for custom network scenarios and high flexibility, but requires more code to be written. 

Reviewer: add a new line

> Running the binary, the user can send requests to the RPC server in order to start a network, create subnets, add nodes to the network, remove nodes from the network, restart nodes, etc.. You can make requests through the `avalanche-network-runner` command or by making API calls. Requests are "translated" into gRPC and sent to the server.

Reviewer: ```suggestion
Running the binary, the user can send requests to the RPC server in order to start a network, create subnets, add nodes to the network, remove nodes from the network, restart nodes, etc. You can make requests through the `avalanche-network-runner` command or by making API calls. Requests are "translated" into gRPC and sent to the server.
```

Each node can then also be reached via [api](https://github.com/ava-labs/avalanche-network-runner/tree/main/api) endpoints which each node exposes.

The following diagram is a simplified view of the high level architecture of the tool:

> ![ANR architecture](/img/grpc-networkrunner.svg "Simplified ANR Architecture")

Reviewer: ```suggestion
![ANR architecture](/img/grpc-networkrunner.svg)
```

## Examples 

When running with the binary, ANR runs a server process as an RPC server which then waits for API calls and handles them.
Therefore we run one shell with the RPC server, and another one for issuing calls.

### Start the server

```bash
avalanche-network-runner server \
--log-level debug \
--port=":8080" \
--grpc-gateway-port=":8081"
```

Note that the above command will run until you stop it with `CTRL + C`. Further commands will have to be run in a separate terminal.

The RPC server listens to two ports:

> * `port`: the main gRPC port (see [gRPC](https://grpc.io/)

Reviewer: ```suggestion
* `port`: the main gRPC port (see [gRPC](https://grpc.io/)).
```

> * `grpc-gateway-port`: the gRPC gateway port (see [gRPC-gateway](https://grpc-ecosystem.github.io/grpc-gateway/) ), which allows for HTTP requests

Reviewer: ```suggestion
* `grpc-gateway-port`: the gRPC gateway port (see [gRPC-gateway](https://grpc-ecosystem.github.io/grpc-gateway/)), which allows for HTTP requests
```

When using the binary to issue calls, the main port will be hit. In this mode, the binary executes compiled code to issue calls.
Alternatively, plain HTTP can be used to issue calls, without the need to use the binary. In this mode, the `grpc-gateway-port` should be queried.

Each of the examples below will show both modes, claritying its usage.

### Run queries

#### ping the server

```bash
curl -X POST -k http://localhost:8081/v1/ping -d ''

# or
avalanche-network-runner ping \
--log-level debug \
--endpoint="0.0.0.0:8080"

> ```

Reviewer: Please use separate ```bash for `curl` and `avalanche-network-runner` so that it is easy to copy+paste the command. This applies to all the commands in this doc.

        "health-check-frequency":"2s",
        "api-admin-enabled":true,
        "api-ipcs-enabled":true,
        "index-enabled":true
        }
    ```
2. `--global-node-config` is a JSON string representing a *single* avalanchego config, which will be applied to **all nodes**. This makes it easy to define common properties to all nodes. Whatever is set here will be *combined* with the standard set above.
3. `--custom-node-configs` is a map of JSON strings representing the *complete* network with individual configs. This allows to configure each node independently. If set, `--number-of-nodes` will be **ignored** to avoid conflicts.
4. The configs can be combined and will be merged, i.e. one could set global `--global-node-config` entries applied to each node, and also set `--custom-node-configs` for additional entries.
5. Common `--custom-node-configs` entries override `--global-node-config` entries which override the standard set.
6. The following entries will be **ignored in all cases** because the network-runner needs to set them internally to function properly:
    ```bash
      --log-dir
      --db-dir
      --http-port
      --staking-port
      --public-ipc
    ```

> #### wait for all the nodes in the cluster to become healthy

Reviewer: What are the response to indicate that the cluster is healthy or not?

Author: Great point! The response is actually huge, but contains `healthy:true` at the very end. Will add accordingly.

# or
avalanche-network-runner control restart-node \
--request-timeout=3m \
--log-level debug \
--endpoint="0.0.0.0:8080" \
--node-name node1 \
--avalanchego-path ${AVALANCHEGO_EXEC_PATH}
```

#### add a node

In this example we are adding a node named `node1`.

```bash
# e.g., ${HOME}/go/src/github.com/ava-labs/avalanchego/build/avalanchego
AVALANCHEGO_EXEC_PATH="avalanchego"

# Note that you can add the new node with a different binary by providing
# a different execPath
curl -X POST -k http://localhost:8081/v1/control/addnode -d '{"name":"node99","execPath":"'${AVALANCHEGO_EXEC_PATH}'","logLevel":"INFO"}'

# or
avalanche-network-runner control add-node \
--request-timeout=3m \
--log-level debug \
--endpoint="0.0.0.0:8080" \
--node-name node99 \
--avalanchego-path ${AVALANCHEGO_EXEC_PATH}
```

It's also possible to provide custom parameters, similar to starting the network:

```bash
	--node-config '{"index-enabled":false, "api-admin-enabled":true,"network-peer-list-gossip-frequency":"300ms"}'
	--custom-vms '{"subnetevm":"/tmp/subnet-evm.genesis.json"}'
```

`--node-config` allows to specify specific avalanchego config parameters to the new node.

> See [here](https://docs.avax.network/build/references/avalanchego-config-flags) for the reference of supported flags.

Reviewer: please use relative doc path, rather hardcoded https://docs.avax.network/ path

**Note**: The following parameters will be *ignored* if set in `--node-config`, because the network runner needs to set its own in order to function properly:
`--log-dir`
`--db-dir`

`--custom-vms` allows to configure custom VMs supported by this node.
See the [subnet](#network-runner-rpc-server-subnet-evm-example) section for details about how to run subnets.

**Note**: The following subnet parameters will be set from the global network configuration to this node:
`--whitelisted-subnets`
`--plugin-dir`

#### terminate the cluster

Note that this will still require to stop your RPC server process with `Ctrl-C` to free the shell.

```bash
curl -X POST -k http://localhost:8081/v1/control/stop -d ''

# or
avalanche-network-runner control stop \
--log-level debug \
--endpoint="0.0.0.0:8080"
```

## Subnets
ANR can be a great helper tool working with subnets. We recommend using it to develop and test new subnets before deploying them in public networks.

> For general subnet documentation, please refer to [subnets](https://docs.avax.network/build/tutorials/platform/subnets/). 

Reviewer: please use relative doc path, rather hardcoded https://docs.avax.network/ path
